File: tools/python/Cibyl/BinaryTranslation/Optimization/Builtins/exceptions.py
######################################################################
##
## Copyright (C) 2007,  Simon Kagstrom
##
## Filename:      exceptions.py
## Author:        Simon Kagstrom <simon.kagstrom@gmail.com>
## Description:   Exception handling (done through builtins)
##
## $Id:$
##
######################################################################
from Cibyl.BinaryTranslation import bytecode, register
from Cibyl.BinaryTranslation.Mips import mips
import builtins
from base import BuiltinBase
from sets import Set

tryInstruction = None

class ExceptionBuiltinBase(BuiltinBase):
    def __init__(self, controller, instruction, name, operation):
        BuiltinBase.__init__(self, controller, instruction, name, operation)
        self.instruction.destinations = Set([mips.R_EAR, mips.R_ECB])
        self.controller.addLabel(self.instruction.address)

class Try(ExceptionBuiltinBase):
    def compile(self):
        global tryInstruction
        self.rh.pushRegister(mips.R_A0)
        self.rh.popToRegister(mips.R_ECB)
        self.rh.pushRegister(mips.R_A1)
        self.rh.popToRegister(mips.R_EAR)
        tryInstruction = self

class Catch(ExceptionBuiltinBase):
    def compile(self):
        javaMethod = self.instruction.getJavaMethod()
        start = tryInstruction.instruction.address
        end = self.instruction.address

        # Get the labels for this exception
        handlerLabel = javaMethod.addExceptionHandler(start, end)
        startLabel = self.controller.getLabel(start)
        endLabel = self.controller.getLabel(end)

        self.bc.emit(".catch all from %s to %s using %s" % (startLabel, endLabel, handlerLabel) )

    def maxOperandStackHeight(self):
        # Conservative estimate: Itself requires a height of 6, but it
        # may be thrown by code which has up to 11 (also pessimistic)
        # entries on the stack
        return 18

def match(controller, instruction, name):
    names = {
	"__NOPH_try": Try,
	"__NOPH_catch": Catch,
	}
    try:
	cur = names[name]
	return cur(controller, instruction, name, None)
    except:
        return None

builtins.alwaysInline.append("__NOPH_try")
builtins.alwaysInline.append("__NOPH_catch")
builtins.addBuiltin(match)

File: tools/compiled/translator/builtins/exceptions.cc
/*********************************************************************
 *
 * Copyright (C) 2008,  Simon Kagstrom
 *
 * Filename:      exceptions.cc
 * Author:        Simon Kagstrom <simon.kagstrom@gmail.com>
 * Description:   Impl. of exception builtins
 *
 * $Id:$
 *
 ********************************************************************/
#include <javamethod.hh>
#include <controller.hh>
#include <builtins.hh>
#include <emit.hh>

class ExceptionBuiltinBase : public Builtin
{
public:
  ExceptionBuiltinBase(const char *name) : Builtin(name)
  {
  }

  bool pass1(Instruction *insn)
  {
    insn->setBranchTarget();
    return true;
  }

  int fillDestinations(int *p)
  {
    return this->addToRegisterUsage(R_EAR, p) + this->addToRegisterUsage(R_ECB, p);
  };
};

class ExceptionBuiltinTry : public ExceptionBuiltinBase
{
public:
  ExceptionBuiltinTry() : ExceptionBuiltinBase("__NOPH_try")
  {
  }

  int fillSources(int *p)
  {
    return this->addToRegisterUsage(R_A0, p) + this->addToRegisterUsage(R_A1, p);
  };

  bool pass2(Instruction *insn)
  {
    emit->bc_pushregister(R_A0);
    emit->bc_popregister(R_ECB);
    emit->bc_pushregister(R_A1);
    emit->bc_popregister(R_EAR);

    controller->pushTryStack(insn);

    return true;
  }
};
